### PR TITLE
DSEGOG-289 DSEGOG-291 DSEGOG-292 DSEGOG-293 DSEGOG-290 Implement functions with Lark

### PR DESCRIPTION
**Pull request by patrick-austin**
## New functionality
- New endpoints:
  - `/functions` for validating types and variable names of a proposed function (intended to be used duringthe popup, before submitting request to `/records`
  - `/functions/tokens` to get the list of supported tokens, "nice" names and implementation details
  - `/images/function` to generate a single full size image from a function
  - `/waveforms/function` to return x and y of a waveform generated from a function
- Altered endpoints:
  - `/records` accepts functions as an optional query param
- Tests for new functionality

## Other changes
- Development done against python 3.11, and have altered dependencies to allow this (see TODOs)
- Use `json.dumps` instead of `str` in `WaveformModel` to properly handle NaNs and Inf values in the response
- Refactor `get_channel_dtype`

## TODOs
- ~Considering refactoring builtin functions (no functional impact). While the current builtins cover everything in the requirements, more might be needed later, and it has been raised that a more "reflexive" approach might be more sustainable. While explicitly defining functions was OK originally (min, max, log etc.) with all the builtins updating all the places they need references can be tedious and easy to miss one. Refactoring onto a single source of truth and using that might be more maintainable in the long run, e.g. having a single `function` rule in the parser and doing lookup in its `Transformer` function based on a single dict/class (to be created).~
  - Implemeted this change, all builtin functions are now defined statically within a class extending the `ABC` `Builtin`. All evaluation functionality is defined here, with shared functionality defined on `Builtin`. Properties are used to specify accepted types and tokens for the frontend. Once a new class is added, it will need to be added to the `dict` on `Builtins` but other than that no changes to the other Lark elements are needed.
  - Note that for operands and functions where we just call `np`, these are still explicitly defined. Could migrate these to the same architecture as the other builtins, but they are a lot simpler and are well defined so it may be overkill.
- ~Properly handle the Python version issues in a separate PR~
- Frontend dev (and feedback for any required backend changes)
- Validation on generated data (gemini data has limited value for "realistic" results...)
  - Do functions give sensible results, in particular smoothing
  - Do we handle pixel depth properly

### CODE REVIEW

**Comment by patrick-austin** [comment]
- [x] Added high level md summary of functions
- [x] Increased detail in `Transformer` docstrings
- [ ] Add examples to postman

**Comment by patrick-austin** [comment]
Updated to work with export functionality from #104 
Test failures will be fixed by #109 and #110

**Review by kevinphippsstfc — CHANGES_REQUESTED** [review]

File: operationsgateway_api/src/functions/builtins/background.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            Builtin.smooth(argument.y)
+            if len(argument.y) < 50:
+                return np.mean(argument.y)
+            else:
+                return (np.mean(argument.y[:25]) + np.mean(argument.y[-25:])) / 2
+        elif isinstance(argument, np.ndarray):
+            return np.mean(argument[:10, :10])
+        else:
+            message = (
+                "'background' accepts [WaveformVariable, np.ndarray], "
```
[the remark this answers is not in the document]
Comment (patrick-austin) [reply operationsgateway_api/src/functions/builtins/background.py:51]: If things work as intended, then the user would see (or at least we'd provide the frontend with) the message generated by:
https://github.com/ral-facilities/operationsgateway-api/blob/035f38b5a033f2af2eba3ed32b7b722f7559eced/operationsgateway_api/src/functions/builtins/builtins.py#L41-L60
As we would type check functions explicitly before evaluation. There's a bit more effort in mapping the type to something human readable (obviously not for the types we accept, but would need to try and map the type actually provided by the user). But can definitely be done like the type check message here as well.
It's also worth mentioning that functions like FWHM won't throw explicit errors here (but would still error if you gave them the wrong type), it's only because `background` and `integrate` can apply to either WF or images that we're doing the type check.

[the remark this answers is not in the document]
Comment (patrick-austin) [reply operationsgateway_api/src/functions/builtins/background.py:51]: Actually this would only happen in places where we call `Background.evaluate` directly, usually it would be done by https://github.com/ral-facilities/operationsgateway-api/blob/035f38b5a033f2af2eba3ed32b7b722f7559eced/operationsgateway_api/src/functions/builtins/builtins.py#L62-L67
Which also does the above, human readable type check.

File: operationsgateway_api/src/records/record.py [hunk 2]
```diff
@@ -386,3 +413,264 @@ async def convert_search_ranges(date_range, shotnum_range):
...
+        Extracts and returns the relevant data from `channel_value`, handling
+        extra calls needed for "image" and "waveform" types.
+        """
+        channel_dtype = await Record.get_channel_dtype(
+            record_id=record_id,
+            channel_name=name,
+            channel_value=channel_value,
+        )
+        if channel_dtype == "image":
+            # Loading from echo takes ~ 0.2 s per image
```
[the remark this answers is not in the document]
Comment (patrick-austin) [reply operationsgateway_api/src/records/record.py:556]: Based on the commit history, I wrote this in January and Matt messaged slack re: the simulated data in February, so I think it must have been the Gemini data. I think it is definitely size dependent when we were testing things around the meeting w/ Stephen for bit depth, but this sort of thing should be rigorously tested as part of https://stfc.atlassian.net/browse/DSEGOG-324 so I'll delete this comment for now.

File: operationsgateway_api/src/records/record.py [hunk 3]
```diff
@@ -386,3 +413,264 @@ async def convert_search_ranges(date_range, shotnum_range):
...
+        # We do not track the bit depth of inputs, so keep maximum depth to
+        # avoid losing information
+        bits_per_pixel = 16
+        if return_thumbnails:
+            metadata = {
+                "channel_dtype": "image",
+                "x_pixel_size": result.shape[1],
+                "y_pixel_size": result.shape[0],
+            }
+            step_size = max(*result.shape) // 50
```
[the remark this answers is not in the document]
Comment (patrick-austin) [reply operationsgateway_api/src/records/record.py:649]: Yeah I likely used the default thumbnail size of 50 and never got round to generalising it to the actual setting.

It's taking the larger of width and height, then doing floor division to get the number of pixels in the original image that need to be mapped to one of the 50 output pixels. Then just use python slicing to downsample our original image. It's far more quick and dirty than using PIL like we do in the actual thumbnail script. Could do that, would probably introduce overhead but it might be small compared to getting the data from echo, haven't tested it (this was just the simplest way of getting a small image out).

File: operationsgateway_api/src/routes/records.py [hunk 4]
```diff
@@ -109,6 +113,16 @@ async def get_records(
             if truncate:
                 Record.truncate_thumbnails(record_data)
 
+        if functions:
```
[the remark this answers is not in the document]
Comment (patrick-austin) [reply operationsgateway_api/src/routes/records.py:116]: I was under the impression that this was purely a feature for ease of development when comparing bytes for tests etc, so hadn't bothered with it but have added it as an argument to the functions functions so we can apply it directly when generating the thumbnail (with the idea that doing it then is slightly more efficient than iterating again later).

**Comment by kevinphippsstfc** [comment]
Whilst testing the export of channels defined by functions I found a problem with exporting a function of a function on an image channel.

I had the following request parameters set:
conditions: {"_id": {"$eq": "20230605100000"}}
projection: channels.CM-202-CVC-CAM-1
functions: {"name": "img_fn1", "expression": "CM-202-CVC-CAM-1 %2A 2"}
projection: channels.img_fn1
functions: {"name": "img_fn2", "expression": "img_fn1 %2A 3"}
projection: channels.img_fn2

If I remove the bottom 2 parameters above it works but once I add them (containing the function of a function) it fails.

At first it seemed the "metadata" for the img_fn1 was not being set but after we set that, it was then trying to download the image images/20230605100000/img_fn1.png from Echo which obviously does not exist.

**Comment by kevinphippsstfc** [comment]
I decided to test the /records endpoint with the same parameters as I tried on the /export endpoint and I got a very similar error:
"Image not available due to incorrect record ID or channel name"
and again it is trying and failing to download the images/20230605100000/img_fn1.png file from Echo.

**Comment by kevinphippsstfc** [comment]
I also tried similar for functions on waveforms on the /records endpoint and the function of a function fails in a similar way to for images.
There is a KeyError for waveform_path on line 569 in _extract_variable when there is not going to be a waveform path for the waveform generated by the function because it doesn't exist in Echo.

I presume that this would also be a problem on the /export endpoint but have not tried that.

I did try doing a function of a function for a scalar channel and that did work, so that's a bonus!

**Comment by patrick-austin** [comment]
> I also tried similar for functions on waveforms on the /records endpoint and the function of a function fails in a similar way to for images. There is a KeyError for waveform_path on line 569 in _extract_variable when there is not going to be a waveform path for the waveform generated by the function because it doesn't exist in Echo.
> 
> I presume that this would also be a problem on the /export endpoint but have not tried that.
> 
> I did try doing a function of a function for a scalar channel and that did work, so that's a bonus!

Problem was that images/waveforms were being added to the record as thumbnail (bytes) or their raw form (bytes/`WaveformModel`), whereas a scalar is just a float. So when stacking functions, it was able to directly extract the scalar but the waveform and image are not in a format that could be directly used by the function (i.e. a np.ndarray).

Added a `_variable_value` key which stores these in the right format and directly extracts this when iterating over the functions. This is then removed before returning the final record.

Added two additional test cases which use image and waveform function outputs to test this.

**Comment by patrick-austin** [comment]
Changes to validation to address edge cases and make error handling in the frontend easier as raised by @joshuadkitenge :

- `/functions/validate` now POST instead of GET to make error handling easier, as a consequence:
  - No longer stops once a certain `name is reached and will evaluate all functions passed and return all data types as a list
  - Accepts a `List[Function]` (a new Pydantic model) as a body, where the `name` and `expression` fields are constrained strings with a minimum length of 1
  - Errors now include the index of the function which caused the error to make associating this to input in frontend easier
- Increased error handling for the `name` (in addition to existing channel name check):
  - Check if another function has already used the same name
  - Check if another builtin has the same name
  - Parse the `name` and check that it is of the `"variable"` type (this will also fail for generic parser errors, caused by e.g. the presence of special characters not expected by the parser like `@`)
- Updates to existing tests and added cases for the new checks described above
- Added detail to the validate docstring to describe how functions depend on each other and are validated in order
- Updated Postman to use the new POST request (body is the same as the old query param example)

**Comment by patrick-austin** [comment]
Improved handling for functions when a channel name exists (in the manifest) but does not have a value defined for a particular record, as raised by @joshuadkitenge :
- Missing channels are removing from the "running" list of variables to store, so that future functions that do not depend on them can still be evaluated.
- Functions that cannot be calculated for the above reason are stored in another `set`, so that we can gracefully skip functions that depend on them (like we would if a channel wasn't set for that record)
- Added a test case to cover both of the above scenarios.

**Review by kevinphippsstfc — APPROVED** [review]

**Comment by patrick-austin** [comment]
Updates to:
- Use the new ChannelManifest object for functions functionality
- Use the renamed config setting for thumbnail size
- Update phash in functions tests for waveforms to account for new config settings for size and linewidth
- poetry.lock

**Review by MRichards99 — APPROVED** [review]
Merge changes look good, Safety CI job also fixed

**Review by MRichards99 — APPROVED** [review]
Looks ready to merge :)
